test/cpp/api/autograd.cpp

```diff
@@ -0,0 +1,129 @@
+#include <gtest/gtest.h>
+
+#include <torch/autograd.h>
+
+#include <torch/utils.h>
+#include <test/cpp/api/support.h>
+
+using namespace torch;
+
+#define ASSERT_VARIABLE_EQ(a,b) ASSERT_TRUE(torch::allclose((a),(b)))
+
+std::string graph_desc(std::shared_ptr<autograd::Node> node) {
+  if (!node) {
+    return "None";
+  }
+  auto result = node->name() + "(";
+  auto next_edges = node->next_edges();
+  for(auto& edge : next_edges) {
+    result += graph_desc(edge.function);
+  }
+  return result+")";
+}
+
+TEST(CustomAutogradTest, CustomFunction) {
+  struct MyFunction : public torch::autograd::Function<MyFunction> {
+    static variable_list forward(AutogradContext *ctx, Variable var1, int mul, Variable var2) {
+      ctx->saved_data["mul"] = mul;
+      ctx->save_for_backward({var1, var2});
+      return {var1 + mul*var2 + var1*var2};
+    }
+
+    static variable_list backward(AutogradContext *ctx, variable_list grad_output) {
+      int mul = ctx->saved_data["mul"].toInt();
+      auto saved = ctx->get_saved_variables();
+      auto var1 = saved[0];
+      auto var2 = saved[1];
+      variable_list output = {grad_output[0] + grad_output[0]*var2, Variable(), grad_output[0] * mul + grad_output[0] * var1};
+      return output;
+    }
+  };
+
+  Variable x = torch::randn({5,5}, torch::requires_grad());
+  Variable y = torch::randn({5,5}, torch::requires_grad());
+  auto res = MyFunction::apply(x,2,y)[0];
```

Comment:
Hmm, I guess we could probably do some shenanigans to make single output functions nicer (extra template parameter to Function which specifies the return type of apply), but that'd be some extra C++ work.

Reply:
Maybe we can file an enhancement request for this and see if someone wants to pick it up.